### ISSUE TITLE
Feature request: set target limits via service discovery labels

### ISSUE THREAD

**Issue by krya-kryak** [prometheus/prometheus#10441]
## Proposal
TL;DR: User should be able to set various scrape-related limits (`body_size_limit`, `sample_limit`, `label_limit`, etc) for individual targets based on its discovered labels.

## Use case
Consider a typical prometheus config:
```
scrape_configs:
- job_name: all-the-targets
  sample_limit: 100
  <some_discovery_config>
```
This is perfectly readable and works just fine up until some special important target appears. This target contains more than 100 important metrics, and setting different `sample_limit` is a necessity. However, you have an understanding that there is still a reasonable upper limit. The config grows:
```
scrape_configs:
- job_name: all-the-targets-but-target1
  sample_limit: 100
  relabel_config:
  - action: drop
    source_labels: [__sd_target_name]
    regex: target1
  <some_discovery_config>
- job_name: target1
  sample_limit: 200
  relabel_config:
  - action: keep
    source_labels: [__sd_target_name]
    regex: target1
  <some_discovery_config>
```
Not so pretty, but still readable. Time passes, and important endpoint appears. It contains even more metrics. Something like this happens:
```
scrape_configs:
- job_name: all-the-targets-but-target1-or-target2
  sample_limit: 100
  relabel_config:
  - action: drop
    source_labels: [__sd_target_name]
    regex: target1|target2
  <some_discovery_config>
- job_name: target1
  sample_limit: 200
  relabel_config:
  - action: keep
    source_labels: [__sd_target_name]
    regex: target1
- job_name: target2
  sample_limit: 10000
  relabel_config:
  - action: keep
    source_labels: [__sd_target_name]
    regex: target2
  <some_discovery_config>
```

Then prometheus 2.27 comes along, and prometheus admin tries to enforce `label_value_length_limit`. However, there are several endpoints that already have weirdly long labels. They should be allowed to function, but no new endpoints should be allowed to behave that way. This is when hell brakes loose:
```
scrape_configs:
- job_name: all-the-targets-but-target1-or-target2-and-longlabeltargets
  sample_limit: 100
  label_value_length_limit: 20
  relabel_config:
  - action: drop
    source_labels: [__sd_target_name]
    regex: target1|target2
  - action: drop
    source_labels: [__sd_target_has_long_labels]
    regex: true
  <some_discovery_config>
- job_name: longlabeltargets
  sample_limit: 100
  label_value_length_limit: 100
  relabel_config:
  - action: keep
    source_labels: [__sd_target_has_long_labels]
    regex: true
  <some_discovery_config>
- job_name: target1
  sample_limit: 200
  label_value_length_limit: 20
  relabel_config:
  - action: keep
    source_labels: [__sd_target_name]
    regex: target1
- job_name: target2
  sample_limit: 10000
  label_value_length_limit: 20
  relabel_config:
  - action: keep
    source_labels: [__sd_target_name]
    regex: target2
  <some_discovery_config>
```
pod_name is used just as an example. various labels would be used in production system, making changes more and more difficult with every change. At some point operator forgets to set some new limit. Those are details. The essence is: config becomes complicated, hard to read and to maintain.

Adjusting some per-target limits via labels will make config dramatically simpler and easier to maintain:
```
scrape_configs:
- job_name: all-the-targets
  sample_limit: 10
  label_value_length_limit: 100
  <some_discovery_config>
```
And some target-specific setting will be set via its labels:
```
[
  {"targets": [ "target1"], "labels": {"__sample_limit__": "200"}},
  {"targets": [ "target2"], "labels": {"__sample_limit__": "10000"}},
  {"targets": [ "longlabeltarget1"], "labels": {"__label_value_length_limit__": ": 100"}},
  {"targets": [ "defaultlimitstarget1"], "labels": {}}
  {"targets": [ "defaultlimitstarget2"], "labels": {}}
 ]
```

Adding a new limit globally will be as easy as it is supposed to be - just add it to a single scrape_config. Adding exceptions to those default limits would only require setting those labels via SD, which could be done without redeploying new prometheus config.

**Similar feature is already present for `interval` and `timeout`:** https://github.com/prometheus/prometheus/issues/8895
I believe this can be done for other scraper / scrape loop limits and `honor*` settings as well: https://github.com/prometheus/prometheus/blob/38ef68e27f565b4183b98d5e080076870f99ac65/scrape/scrape.go#L540-L550

**Comment by krya-kryak** [comment]
A WIP PR to show this is feasible: https://github.com/prometheus/prometheus/pull/10432

**Comment by krajorama** [comment]
Hello from the bug scrub,

closing the issue as there was no further progress after the initial POC. Anyone who would like to work on this, please follow up.

**Comment by beorn7** [comment]
Just in case it helps, my personal gut feeling is we shouldn't do this.

It's very tempting to mirror more and more parts of the scrape config in labels. However, in general there should be one way of doing something. The parts we allowed as labels so far were justified by some notion that targets should have some control of how they wish to be scraped, e.g. if a target "knows" it exposes metrics that change often and make sense to be scraped with high frequency, it makes sense to communicate to Prometheus servers (and possibly other scrapers) via SD that this target wishes to be scraped every second.

However, target limits are inherently limits of the scraper. If a target wishes to limit the metrics it exposes, it can just do so. It doesn't have to be communicated via SD.
